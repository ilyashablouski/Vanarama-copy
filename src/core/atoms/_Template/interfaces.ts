import { IBaseProps } from '../../interfaces/base';

export interface ITemplateProps extends IBaseProps {}
