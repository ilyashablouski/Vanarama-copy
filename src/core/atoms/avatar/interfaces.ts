import { IBaseProps } from '../../interfaces/base';
import { TSize } from '../../../types/size';

export interface IAvatarProps extends IBaseProps {
  src?: string;
  size?: TSize;
  altText?: string;
}
