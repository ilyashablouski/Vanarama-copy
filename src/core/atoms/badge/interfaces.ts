import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IBadgeProps extends IBaseProps {
  size?: TSize;
  color?: TColor;
  label?: string;
}
