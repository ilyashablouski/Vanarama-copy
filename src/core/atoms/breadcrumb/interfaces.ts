import { IBaseProps } from '../../interfaces/base';
import { ILink } from '../../interfaces/link';

export interface IBreadcrumbProps extends IBaseProps {
  items?: ILink[];
}
