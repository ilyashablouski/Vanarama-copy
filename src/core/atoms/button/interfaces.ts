import { ButtonHTMLAttributes } from 'react';
import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';
import { TSize } from '../../../types/size';

export interface IButtonProps
  extends IBaseProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  color?: TColor;
  fill?: 'solid' | 'outline' | 'clear';
  icon?: React.ReactNode;
  iconColor?: TColor;
  iconPosition?: 'before' | 'after';
  label?: string | React.ReactNode;
  round?: boolean;
  size?: TSize;
  withoutDefaultClass?: boolean;
  customCTAColor?: string;
}
