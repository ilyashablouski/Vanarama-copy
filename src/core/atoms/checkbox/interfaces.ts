import { InputHTMLAttributes, ReactElement } from 'react';
import { IBaseProps } from '../../interfaces/base';
import { TColor } from '../../../types/color';

export interface ICheckboxProps
  extends IBaseProps,
    InputHTMLAttributes<HTMLInputElement> {
  /**
   * A color for the checkbox.
   */
  color?: TColor;
  /**
   * A unique ID for the checkbox node - this is required here because the
   * underlying checkbox is hidden and therefore you can only click on the label
   * and the label needs to know which input it corresponds to.
   */
  id: string;
  /**
   * 	Invalid state of the checkbox.
   */
  invalid?: boolean;
  /**
   * A name for the input node.
   */
  label: string | (string | ReactElement)[];
  /**
   * Outlined state of the checkbox.
   */
  outline?: boolean;
  /**
   * A ref to attach to the checkbox.
   */
  ref?: React.Ref<HTMLInputElement>;
}
