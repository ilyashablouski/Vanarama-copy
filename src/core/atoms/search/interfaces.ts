import { IBaseProps } from '../../interfaces/base';

export interface ISearchProps extends IBaseProps {
  id?: string;
  name?: string;
  placeholder?: string;
  results?: string[];
  onSelect?: (value: string) => void;
  onChange?(event: React.ChangeEvent<HTMLInputElement>): void;
  isIntermediateState?: boolean;
}
