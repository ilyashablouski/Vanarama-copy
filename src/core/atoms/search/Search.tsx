import cx from 'classnames';
import React, { FC, memo, useRef, useState } from 'react';
import useDeepCompareEffect from 'use-deep-compare-effect';
import Close from '../../assets/icons/Close';
import SearchIcon from '../../assets/icons/Search';
import Tile from '../../molecules/tile';
import Button from '../button';
import Icon from '../icon';
import { ISearchProps } from './interfaces';
import Text from '../text';
import Link from '../link';

const Search: FC<ISearchProps> = memo(props => {
  const {
    id,
    name,
    className,
    placeholder,
    results = [],
    onSelect,
    onChange,
    isIntermediateState,
  } = props;

  const [searchValue, setSearchValue] = useState<string>('');
  const [searchResults, setSearchResults] = useState<string[]>([]);
  // const [cursor, setCursor] = useState(0);

  const inputRef = useRef<HTMLInputElement>(null);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
    if (onChange) {
      onChange(e);
    }
  };

  const handleSelect = (result: string) => {
    setSearchValue(result.trim());
    if (inputRef.current) {
      inputRef.current.focus();
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === 13) {
      setSearchResults([]);
      if (onSelect) {
        onSelect(searchValue);
      }
    }
  };

  // >>> reset search results <<<
  const handleClear = () => {
    setSearchValue('');
    setSearchResults([]);
  };

  useDeepCompareEffect(() => {
    const newResults = results.filter(result => {
      return result.toLowerCase().includes(searchValue.toLowerCase());
    });
    setSearchResults(newResults);
  }, [results, searchValue]);

  return (
    <div className={cx('search--wrapper', className)}>
      <div className={cx('search', { '-results': results.length > 0 })}>
        <Icon icon={<SearchIcon />} size="regular" className="magnifier" />
        <label className="search-label" />
        <input
          id={id}
          name={name}
          ref={inputRef}
          role="searchbox"
          className="search-input"
          type="text"
          value={searchValue}
          placeholder={placeholder}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
        />
        <Button
          type="button"
          label={<Icon icon={<Close />} size="small" color="success" />}
          fill="clear"
          className="search-close"
          onClick={() => handleClear()}
        />
      </div>
      {isIntermediateState && (
        <div className="search--spacer">
          <Text>
            Velit dolore exercitationem distinctio ipsa numquam quisquam?
          </Text>
          <Link>Change</Link>
        </div>
      )}
      {searchResults.length > 0 && (
        <Tile scrollable className="search--results">
          <ul className="search--results--ul">
            {searchResults.map(result => {
              return (
                <li
                  style={{ cursor: 'pointer' }}
                  onClick={() => handleSelect(result)}
                  key={result}
                >
                  <span>{result}</span>
                </li>
              );
            })}
          </ul>
        </Tile>
      )}
    </div>
  );
});

export default Search;
