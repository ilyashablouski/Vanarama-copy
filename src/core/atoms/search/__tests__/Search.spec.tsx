import * as React from 'react';
import { render } from '@testing-library/react';
import Search from '../Search';

import { ISearchProps } from '../interfaces';

const getComponent = (props: ISearchProps) => <Search {...props} />;

describe('<Search/>', () => {
  it('renders default state correctly', () => {
    const { container } = render(getComponent({}));
    expect(container).toMatchSnapshot();
  });

  it('renders Intermediate state correctly ', () => {
    const { container } = render(getComponent({ isIntermediateState: true }));
    expect(container).toMatchSnapshot();
  });
  // TODO: fix test
  it('should contain defined handler and id props on input el', () => {
    // const handler = jest.fn();
    //
    // const props = render(
    //   getComponent({
    //     id: 'testInput',
    //     onChange: handler,
    //   }),
    // )
    //   .getElementsByClassName('search-input')
    //   .first()
    //   .props();
    //
    // expect(props.onChange).toEqual(expect.any(Function));
    // expect(props.id).toEqual('testInput');
  });
  // TODO: fix test
  it('should pass event object with value to handler onChange', () => {
    // const handler = jest.fn();
    // const { container } = render(getComponent({ onChange: handler }));
    // const event = { target: { value: 'newVal' } };
    //
    // container.querySelector('input')?.onchange?.(event as any);
    //
    // expect(handler).toBeCalledWith(
    //   expect.objectContaining({ target: { value: 'newVal' } }),
    // );
  });
});
