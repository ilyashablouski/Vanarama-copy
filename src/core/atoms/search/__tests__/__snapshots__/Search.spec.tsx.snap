// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/> renders Intermediate state correctly  1`] = `
<div
  className="search--wrapper"
>
  <div
    className="search"
  >
    <Icon
      className="magnifier"
      icon={<Memo />}
      size="regular"
    />
    <label
      className="search-label"
    />
    <input
      className="search-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      role="searchbox"
      type="text"
      value=""
    />
    <Memo(ForwardRef(Button))
      className="search-close"
      fill="clear"
      label={
        <Memo(Icon)
          color="success"
          icon={<Memo />}
          size="small"
        />
      }
      onClick={[Function]}
      type="button"
    />
  </div>
  <div
    className="search--spacer"
  >
    <Memo(Text)>
      Velit dolore exercitationem distinctio ipsa numquam quisquam?
    </Memo(Text)>
    <Link>
      Change
    </Link>
  </div>
</div>
`;

exports[`<Search/> renders default state correctly 1`] = `
<div
  className="search--wrapper"
>
  <div
    className="search"
  >
    <Icon
      className="magnifier"
      icon={<Memo />}
      size="regular"
    />
    <label
      className="search-label"
    />
    <input
      className="search-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      role="searchbox"
      type="text"
      value=""
    />
    <Memo(ForwardRef(Button))
      className="search-close"
      fill="clear"
      label={
        <Memo(Icon)
          color="success"
          icon={<Memo />}
          size="small"
        />
      }
      onClick={[Function]}
      type="button"
    />
  </div>
</div>
`;
